File: cloudcubes/cloudcubes_stack.py
import aws_cdk.core as cdk
import aws_cdk.aws_s3 as s3
import aws_cdk.aws_lambda as lambda_
import aws_cdk.aws_iam as iam
import aws_cdk.aws_s3_deployment as s3_deployment

class CloudcubesStack(cdk.Stack):

    def __init__(self, scope: cdk.Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Name of the DynamoDB database
        db_name = cdk.CfnParameter(self, "Database_Name",
            type="String",
            description="The name of the dynamodb database where the information on server data is stored."
        )

        # Grants access to the dynamodb database
        database_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=[f'arn:aws:dynamodb:*:*:table/{db_name.value_as_string}'],
            actions=[
                'dynamodb:BatchGetItem',
                'dynamodb:BatchWriteItem',
                'dynamodb:ConditionCheckItem',
                'dynamodb:PutItem',
                'dynamodb:DescribeTable',
                'dynamodb:DeleteItem',
                'dynamodb:GetItem',
                'dynamodb:Scan',
                'dynamodb:Query',
                'dynamodb:UpdateItem'
            ]
        )

        # The lambda function in charge of scheduling servers
        scheduler_function = lambda_.Function(self, "SchedulerFunction",
            runtime=lambda_.Runtime.PYTHON_3_8,
            code=lambda_.Code.from_asset("./scheduler"),
            handler="app.lambda_handler",
            environment={
                "DATABASE_NAME": db_name.value_as_string
            }
        )
        scheduler_function.add_to_role_policy(database_perms)

        # S3 bucket to store scripts (used by EC2 instances)
        scripts_bucket = s3.Bucket(self, "ScriptsBucket",
            versioned=False,
            removal_policy=cdk.RemovalPolicy.DESTROY
        )

        scripts_bucket_deployment = s3_deployment.BucketDeployment(self, "ScriptsBucketDeployment",
            destination_bucket=scripts_bucket,
            sources=[s3_deployment.Source.asset('./scripts')]
        )

        # Permissions to read scripts bucket
        scripts_bucket_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=[scripts_bucket.bucket_arn, f'{scripts_bucket.bucket_arn}/*'],
            actions=[
                's3:ListBucket',
                's3:GetObject'
            ]
        )

        attach_volumes_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=['*'],
            actions=[
                'ec2:AttachVolume'
            ]
        )

        # Role and InstanceProfile to be assigned to launched EC2 instances
        server_role = iam.Role(self, "Role",
            assumed_by=iam.ServicePrincipal('ec2.amazonaws.com')
        )
        server_role.add_to_policy(database_perms)
        server_role.add_to_policy(scripts_bucket_perms)
        server_role.add_to_policy(attach_volumes_perms)
        server_role.add_managed_policy(iam.ManagedPolicy.from_aws_managed_policy_name('AmazonSSMManagedInstanceCore'))
        server_instance_profile = iam.CfnInstanceProfile(self, "ServerInstanceProfile",
            roles=[server_role.role_name]
        )
        
        # Permissions to start servers
        start_server_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=['*'],
            actions=[
                'ec2:RequestSpotInstances'
            ]
        )
        pass_server_role_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=[server_role.role_arn],
            actions=[
                'iam:GetRole',
                'iam:PassRole'
            ]
        )

        # Lambda function for starting servers
        server_starter_function = lambda_.Function(self, "ServerStarterFunction",
            runtime=lambda_.Runtime.PYTHON_3_8,
            code=lambda_.Code.from_asset('./server-starter'),
            handler="app.lambda_handler",
            environment={
                "DATABASE_NAME": db_name.value_as_string,
                "SCRIPTS_BUCKET": scripts_bucket.bucket_name,
                "SERVER_INSTANCE_PROFILE": server_instance_profile.attr_arn
            }
        )
        server_starter_function.add_to_role_policy(database_perms)
        server_starter_function.add_to_role_policy(start_server_perms)
        server_starter_function.add_to_role_policy(pass_server_role_perms)

        list_ssm_actions_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=['*'],
            actions=[
                'ssm:ListDocuments',
                "ssm:DescribeDocument",
                "ssm:GetDocument",
                "ssm:DescribeInstance*"
            ]
        )
        
        run_command_perms = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            resources=[
                '*'
            ],
            actions=[
                '*'
            ]
        )

        # Lambda function for stopping servers
        server_stopper_function = lambda_.Function(self, "ServerStopperFunction",
            runtime=lambda_.Runtime.PYTHON_3_8,
            code=lambda_.Code.from_asset('./server-stopper'),
            handler="app.lambda_handler",
            environment={
                "DATABASE_NAME": db_name.value_as_string
            }
        )
        server_stopper_function.add_to_role_policy(database_perms)
        server_stopper_function.add_to_role_policy(list_ssm_actions_perms)
        server_stopper_function.add_to_role_policy(run_command_perms)

File: infrastructure/src/main/java/osbourn/cloudcubes/infrastructure/CloudCubesStack.java
package osbourn.cloudcubes.infrastructure;

import osbourn.cloudcubes.core.constructs.InfrastructureConfiguration;
import osbourn.cloudcubes.core.constructs.InfrastructureConfiguration.InfrastructureSetting;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.RemovalPolicy;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.services.dynamodb.Attribute;
import software.amazon.awscdk.services.dynamodb.AttributeType;
import software.amazon.awscdk.services.dynamodb.BillingMode;
import software.amazon.awscdk.services.dynamodb.Table;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.iam.*;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.Runtime;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.deployment.BucketDeployment;
import software.amazon.awscdk.services.s3.deployment.Source;
import software.constructs.Construct;

import java.util.*;

public class CloudCubesStack extends Stack {
    public CloudCubesStack(final Construct parent, final String name) {
        super(parent, name);

        // Create the DynamoDB table that stores the data for the servers
        Attribute serverTablePartitionKey = Attribute.builder()
                .name("Id")
                .type(AttributeType.NUMBER)
                .build();
        Table serverTable = Table.Builder.create(this, "ServerTable")
                .removalPolicy(RemovalPolicy.RETAIN)
                .billingMode(BillingMode.PAY_PER_REQUEST)
                .partitionKey(serverTablePartitionKey)
                .build();

        // Resources bucket: the contents of the resources folder will be made available as an S3 bucket
        Bucket resourceBucket = Bucket.Builder.create(this, "ResourceBucket")
                .removalPolicy(RemovalPolicy.DESTROY)
                .autoDeleteObjects(true)
                .build();
        BucketDeployment resourceBucketDeployment = BucketDeployment.Builder.create(this, "ResourceBucketDeployment")
                .destinationBucket(resourceBucket)
                .sources(Collections.singletonList(Source.asset("./resources")))
                .build();

        // Create VPC
        Vpc serverVpc = Vpc.Builder.create(this, "ServerVpc")
                // This will force AWS to create public subnets instead of private subnets
                // Without this setting, AWS will create a NAT Gateway for the private subnets, costing $0.045 per hour
                .natGateways(0)
                .build();

        // Create security group used by server instances
        final int sshPort = 22;
        final int minecraftPort = 25565;
        SecurityGroup serverSecurityGroup = SecurityGroup.Builder.create(this, "ServerSecurityGroup")
                .description("Security group for EC2 Instances launched by the CloudCubes application")
                .vpc(serverVpc)
                .allowAllOutbound(true)
                .build();
        Connections connections = serverSecurityGroup.getConnections();
        connections.allowFromAnyIpv4(Port.tcp(minecraftPort), "Allow TCP access to the Minecraft Server");
        connections.allowFromAnyIpv4(Port.udp(minecraftPort), "Allow UDP access to the Minecraft Server");
        connections.allowFromAnyIpv4(Port.tcp(sshPort), "Allows TCP access through SSH");

        // IAM Role for EC2 instances
        Role serverRole = Role.Builder.create(this, "ServerRole")
                .description("Grants EC2 instances permission to access resources from the Cloudcubes stack")
                .assumedBy(ServicePrincipal.Builder.create("ec2.amazonaws.com").build())
                .build();
        serverRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore"));
        serverTable.grantReadWriteData(serverRole);
        resourceBucket.grantRead(serverRole);
        CfnInstanceProfile serverInstanceProfile = CfnInstanceProfile.Builder.create(this, "ServerInstanceProfile")
                .roles(Collections.singletonList(serverRole.getRoleName()))
                .build();

        // Create InfrastructureConfiguration object to determine environment variables for the lambda functions
        List<String> serverSubnetIds = new ArrayList<>();
        for (ISubnet subnet : serverVpc.getPublicSubnets()) {
            serverSubnetIds.add(subnet.getSubnetId());
        }
        InfrastructureConfiguration ic = new InfrastructureConfiguration();
        ic.setValue(InfrastructureSetting.REGIONASSTRING, "US-EAST-2");
        ic.setValue(InfrastructureSetting.SERVERDATABASENAME, serverTable.getTableName());
        ic.setValue(InfrastructureSetting.RESOURCEBUCKETNAME, resourceBucket.getBucketName());
        ic.setValue(InfrastructureSetting.SERVERROLEID, serverRole.getRoleId());
        ic.setValue(InfrastructureSetting.SERVERINSTANCEPROFILEARN, serverInstanceProfile.getAttrArn());
        ic.setValue(InfrastructureSetting.SERVERSECURITYGROUPID, serverSecurityGroup.getSecurityGroupId());
        ic.setValue(InfrastructureSetting.SERVERVPCID, serverVpc.getVpcId());
        ic.setServerSubnetIds(serverSubnetIds);

        Map<String, String> infrastructureDataMap = ic.toEnvironmentVariableMap();

        // Create the server starter function
        Function serverStarter = Function.Builder.create(this, "ServerStarter")
                .code(Code.fromAsset("lambda/server-starter/build/libs/server-starter-all.jar"))
                .handler("osbourn.cloudcubes.lambda.serverstarter.ServerStarterLambdaHandler")
                .runtime(Runtime.JAVA_11)
                .environment(infrastructureDataMap)
                .timeout(Duration.seconds(30))
                .memorySize(512)
                .build();
        assert serverStarter.getRole() != null;
        serverStarter.getRole().addToPrincipalPolicy(PolicyStatement.Builder.create()
                .effect(Effect.ALLOW)
                .resources(Collections.singletonList("*"))
                .actions(Collections.singletonList("ec2:RequestSpotInstances"))
                .build());
        // serverStarter needs a special permission in order to launch servers with IAM roles
        serverStarter.getRole().addToPrincipalPolicy(PolicyStatement.Builder.create()
                .effect(Effect.ALLOW)
                .resources(Collections.singletonList(serverRole.getRoleArn()))
                .actions(Arrays.asList("iam:GetRole", "iam:PassRole"))
                .build());
        serverTable.grantReadWriteData(serverStarter);
    }
}
